hc_dpki/src/keypair.rs

```diff
@@ -0,0 +1,562 @@
+                    decrypted_data[0]
+                )));
+            }
+            sign_priv.write(0, &decrypted_data[65..129])?;
+            enc_priv.write(0, &decrypted_data[129..161])?;
+
+            KeyBuffer::with_raw_parts(
+                array_ref![&decrypted_data, 1, 32],
+                array_ref![&decrypted_data, 33, 32],
+            )
+            .render()
+        };
+
+        Ok(Keypair {
+            pub_keys,
+            enc_priv,
+            sign_priv,
+        })
+    }
+
+    /// sign some arbitrary data with the signing private key
+    ///
+    /// @param {SecBuf} data - the data to sign
+    ///
+    /// @param {SecBuf} signature - Empty Buf the sign
+    pub fn sign(
+        &mut self,
+        data: &mut SecBuf,
+        signature: &mut SecBuf,
+    ) -> Result<(), HolochainError> {
+        sign::sign(data, &mut self.sign_priv, signature)?;
+        Ok(())
+    }
+
+    /// verify data that was signed with our private signing key
+    ///
+    /// @param {SecBuf} signature
+    ///
+    /// @param {SecBuf} data
+    pub fn verify(
+        &mut self,
+        signature: &mut SecBuf,
+        data: &mut SecBuf,
+    ) -> Result<i32, HolochainError> {
+        let mut sign_pub = SecBuf::with_insecure(sign::PUBLICKEYBYTES);
+        let mut enc_pub = SecBuf::with_insecure(kx::PUBLICKEYBYTES);
+
+        util::decode_id(self.pub_keys.clone(), &mut sign_pub, &mut enc_pub)?;
+        let v: i32 = sign::verify(signature, data, &mut sign_pub);
+        Ok(v)
+    }
+
+    /// encrypt arbitrary data to be readale by potentially multiple recipients
+    ///
+    /// @param {array<string>} recipientIds - multiple recipient identifier strings
+    ///
+    /// @param {Buffer} data - the data to encrypt
+    ///
+    /// @param {Buffer} out - Empty vec[secBuf]
+    pub fn encrypt(
+        &mut self,
+        recipient_id: Vec<&String>,
+        data: &mut SecBuf,
+        out: &mut Vec<SecBuf>,
+    ) -> Result<(), HolochainError> {
+        let mut sym_secret = SecBuf::with_secure(32);
+        random_secbuf(&mut sym_secret);
+
+        let mut srv_rx = SecBuf::with_insecure(kx::SESSIONKEYBYTES);
+        let mut srv_tx = SecBuf::with_insecure(kx::SESSIONKEYBYTES);
+
+        let pub_keys = &mut self.pub_keys;
+        let mut sign_pub = SecBuf::with_insecure(sign::PUBLICKEYBYTES);
+        let mut enc_pub = SecBuf::with_insecure(kx::PUBLICKEYBYTES);
+        util::decode_id(pub_keys.to_string(), &mut sign_pub, &mut enc_pub)?;
+
+        for client_pk in recipient_id {
+            let mut r_sign_pub = SecBuf::with_insecure(sign::PUBLICKEYBYTES);
+            let mut r_enc_pub = SecBuf::with_insecure(kx::PUBLICKEYBYTES);
+
+            util::decode_id(client_pk.to_string(), &mut r_sign_pub, &mut r_enc_pub)?;
+
+            kx::server_session(
+                &mut enc_pub,
+                &mut self.enc_priv,
+                &mut r_enc_pub,
+                &mut srv_rx,
+                &mut srv_tx,
+            )?;
```

Comment:
What is this doing and why does it need the encryption private key?

Reply:
The encryption scheme is diffie-hellman based. both parties use their own public/private pair and the counter-parties public half to derive shared symmetric encryption keys. We're not ready for encryption yet, this is a POC that may change as we sort out our requirements.